Sources/BSON/BSONDocument.swift

```diff
@@ -0,0 +1,240 @@
+internal let BSON_ALLOCATOR = ByteBufferAllocator()
+/// Maximum BSON document size in bytes
+internal let BSON_MAX_SIZE = 16_000_000
+/// Minimum BSON document size in bytes
+internal let BSON_MIN_SIZE = 5
+
+/// A struct representing the BSON document type.
+@dynamicMemberLookup
+public struct BSONDocument {
+    /// The element type of a document: a tuple containing an individual key-value pair.
+    public typealias KeyValuePair = (key: String, value: BSON)
+
+    private var _buffer: ByteBuffer
+
+    internal var size: Int {
+        guard let size = self._buffer.getInteger(at: 0, endianness: .little, as: Int32.self) else {
+            fatalError("Cannot read size of BSON from buffer")
+        }
+        return Int(size)
+    }
+
+    /// An unordered set containing the keys in this document.
+    private var keySet: Set<String>
+
+    internal init(_ elements: [BSON]) { fatalError("Unimplemented") }
+
+    internal init(keyValuePairs: [(String, BSON)]) {
+        self.keySet = Set(keyValuePairs.map { $0.0 })
+        guard self.keySet.count == keyValuePairs.count else {
+            fatalError("Dictionary \(keyValuePairs) contains duplicate keys")
+        }
+
+        self._buffer = BSON_ALLOCATOR.buffer(capacity: 100)
+
+        guard !self.keySet.isEmpty else {
+            self = BSONDocument()
+            return
+        }
+
+        let start = self._buffer.writerIndex
+
+        // reserve space for our size that will be calculated
+        self._buffer.writeInteger(0, endianness: .little, as: Int32.self)
+
+        for (key, value) in keyValuePairs {
+            self._buffer.writeInteger(value.bsonValue.bsonType.rawValue, as: UInt8.self)
+            self._buffer.writeCString(key)
+            value.bsonValue.write(to: &self._buffer)
+        }
+        self._buffer.writeInteger(0, as: UInt8.self)
+
+        guard let size = Int32(exactly: self._buffer.writerIndex - start) else {
+            fatalError("Data is \(self._buffer.writerIndex - start) bytes, "
+                + "but maximum allowed BSON document size is \(Int32.max) bytes")
+        }
+        // BSON null terminator
+        self._buffer.setInteger(size, at: 0, endianness: .little, as: Int32.self)
+    }
+
+    /// Initializes a new, empty `BSONDocument`.
+    public init() {
+        self.keySet = Set()
+        self._buffer = BSON_ALLOCATOR.buffer(capacity: 5)
+        self._buffer.writeInteger(5, endianness: .little, as: Int32.self)
+        self._buffer.writeBytes([0])
+    }
+
+    /**
+     * Initializes a new `BSONDocument` from the provided BSON data.
+     *
+     * - Throws:
+     *   - `InvalidArgumentError` if the data passed is invalid BSON
+     *
+     * - SeeAlso: http://bsonspec.org/
+     */
+    public init(fromBSON bson: Data) throws {
+        guard BSONDocument.isValidBSON(bson) else {
+            throw BSONError.InvalidArgumentError(message: "Found `\([UInt8](bson))` is invalid BSON")
```

Comment:
I'm not sure how helpful printing out the byte array is just because it might be very lengthy. 
I'm also thinking it might be better if we just call the method like `validate` and have it throw if the data is not valid. then, depending on why the BSON is invalid, we can provide a more specific error message.